python/ql/test/experimental/dataflow/coverage/test.py

```diff
@@ -39,93 +42,241 @@ def test_floatnumber_literal():
+    for x in gen(SOURCE):
+        SINK(x) # Flow missing
+
 # 6.2.9.1. Generator-iterator methods
-# 6.2.9.2. Examples
+def test___next__():
+    g = gen(SOURCE)
+    SINK(g.__next__()) # Flow missing
+
+def gen2(x):
+    m = yield x # argument of `send` has to flow to value of `yield x` (and so to `m`)
+    yield m
+
+def test_send():
+    g = gen2(3)
+    n = next(g)
+    SINK(g.send(SOURCE)) # Flow missing
+
+def gen_ex(x):
+    try:
+        yield 3
+    except:
+        yield x # `x` has to flow to call to `throw`
+
+def test_throw():
+    g = gen_ex(SOURCE)
+    n = next(g)
+    SINK(g.throw(TypeError)) # Flow missing
+
+# no `test_close` as `close` involves no data flow
+
 # 6.2.9.3. Asynchronous generator functions
+async def agen(x):
+    yield x
+
 # 6.2.9.4. Asynchronous generator-iterator methods
-# 6.3. Primaries
+
+# helper to run async test functions
+def runa(a):
+    import asyncio
+    asyncio.run(a)
+
+async def atest___anext__():
+    g = agen(SOURCE)
+    SINK(await g.__anext__()) # Flow missing
+
+def test___anext__():
+    runa(atest___anext__())
+
+async def agen2(x):
+    m = yield x # argument of `send` has to flow to value of `yield x` (and so to `m`)
+    yield m
+
+async def atest_asend():
+    g = agen2(3)
+    n = await g.__anext__()
+    SINK(await g.asend(SOURCE)) # Flow missing
+
+def test_asend():
+    runa(atest_asend())
+
+async def agen_ex(x):
+    try:
+        yield 3
+    except:
+        yield x # `x` has to flow to call to `athrow`
+
+async def atest_athrow():
+    g = agen_ex(SOURCE)
+    n = await g.__anext__()
+    SINK(await g.athrow(TypeError)) # Flow missing
+
+def test_athrow():
+    runa(atest_athrow())
+
 # 6.3.1. Attribute references
+class C:
+    a = SOURCE
+
+def test_attribute_reference():
+    SINK(C.a) # Flow missing
+
+# overriding __getattr__ should be tested by the class coverage tests
+
 # 6.3.2. Subscriptions
+# This does not constitute dataflow (but could be taint flow)
+def example_subscription_string():
+    SINK("source"[0]) # Flow not expected
+
+def test_subscription_tuple():
+    SINK((SOURCE,)[0]) # Flow missing
+
+def test_subscription_list():
+    SINK([SOURCE][0]) # Flow missing
+
+def test_subscription_mapping():
+    SINK({"s":SOURCE}["s"]) # Flow missing
+
+# overriding __getitem__ should be tested by the class coverage tests
+
 # 6.3.3. Slicings
+l = [SOURCE]
+
+def test_slicing():
+    s = l[0:1:1]
+    SINK(s[0]) # Flow missing
+
+# The grammar seems to allow `l[0:1:1, 0:1]`, but the interpreter does not like it
+
 # 6.3.4. Calls
-# 6.4. Await expression
-# 6.5. The power operator
-# 6.6. Unary arithmetic and bitwise operations
-# 6.7. Binary arithmetic operations
-# 6.8. Shifting operations
-# 6.9. Binary bitwise operations
-# 6.10. Comparisons
-# 6.10.1. Value comparisons
-# 6.10.2. Membership test operations
-# 6.10.3. Identity comparisons
-# 6.11. Boolean operations
+def f(a, b):
+    return b
+
+def test_call_positional():
+    SINK(f(3, SOURCE))
```

Comment:
Also check that

```
SINK(f(SOURCE, NONSOURCE)) # No flow
```

does not result in flow.

Reply:
Good point.

Reply:
Done